### PR TITLE
Fix attn mask ignore logic in training-time trace

### PR DESCRIPTION
**Pull request by zhenglongjiepheonix**
This pr fixes a scenario where we want to use dynamo trace in training mode, the current attn mask ignore logic creates a problem where data-dependent branch condition `torch.all(attn_mask==1)` will cause graph breaks and disable full-graph tracing, the current solution is to disable mask ignore logic as long as we are in tracing mode no matter we are in training or inference phase.

This will enable compilation for training(forward+backward) like this:
```
model = LlamaForCausalLM(config).cuda()
model = torch.compile(model, fullgraph=True)
loss = model(**inputs)[0]
loss.backward()
```

### CODE REVIEW

**Review by ArthurZucker — COMMENTED** [review]
Sounds good! similarly to https://github.com/huggingface/transformers/blob/209fcccbba41575fd0ece8b3ff20930dbac34a80/tests/test_modeling_common.py#L4679 can you add the training compile test? 🤗

**Comment by zhenglongjiepheonix** [comment]
currently cudagraphs are skipped for backward if use `sdpa` and I can't fix it now, I believe there are some attn_mask related bugs in AOT autograd inside `sdpa` because if I replace `scaled_dot_product_attention` with eager forward logic then everything works fine, and the exact problem is the memory of attn mask after processed by `sdpa` traced by AOT autograd is overlapped, which disables cudagraph

**Comment by zhenglongjiepheonix** [comment]
no idea why I am getting this weired `can't find GemmaForCausalLM` issue in CI, can't produce it locally @ArthurZucker 
```
ERROR tests/models/nemotron/test_modeling_nemotron.py - NameError: name 'GemmaForCausalLM' is not defined
ERROR tests/models/nemotron/test_modeling_nemotron.py - NameError: name 'GemmaForCausalLM' is not defined
ERROR tests/models/nemotron/test_modeling_nemotron.py - NameError: name 'GemmaForCausalLM' is not defined
ERROR tests/models/nemotron/test_modeling_nemotron.py - NameError: name 'GemmaForCausalLM' is not defined
```

File: src/transformers/modeling_attn_mask_utils.py [hunk 1]
```diff
@@ -276,7 +276,7 @@ def _ignore_causal_mask_sdpa(
         elif sliding_window is None or key_value_length < sliding_window:
             if len(attention_mask.shape) == 4:
                 return False
-            elif (is_training or not is_tracing) and torch.all(attention_mask == 1):
+            elif not is_tracing and torch.all(attention_mask == 1):
```
Comment (zhenglongjiepheonix) [inline src/transformers/modeling_attn_mask_utils.py:279]: It's actually a paradox branch where `is_tracing` and `torch.all(attention_mask == 1)` can never exist together

File: tests/test_modeling_common.py [hunk 2]
```diff
@@ -4763,6 +4763,52 @@ def test_torch_compile(self):
...
+        batch_size = 1
+        seq_len = 10
+
+        model = cls(config).to(torch_device)
+
+        inputs = {
+            "input_ids": torch.randint(
+                low=1, high=model.config.vocab_size, size=(batch_size, seq_len), device=torch_device
+            ),
+            "attention_mask": torch.ones((batch_size, seq_len), dtype=torch.int64, device=torch_device),
```
Comment (ArthurZucker) [inline tests/test_modeling_common.py:4786]: attn mask being full of one is more prone to skipping some branches, would try with ones and zeros as welll!

**Review by ArthurZucker — APPROVED** [review]
Nice, can you try to run the slow tests for this! 🤗

**Comment by ArthurZucker** [comment]
Failure is most probably not related to you, but the slow tests is bad / badly designed as it should either not use accelerate, or not use parallelism to make sure we are testing apples to apples.

**Comment by ArthurZucker** [comment]
Would be nice if you can update the tests 🙏🏻

**Comment by zhenglongjiepheonix** [comment]
I think previous tests failures will cause gpu memory not release when the compilation tests are executed, making accelerate to kick in. I think the best solution is just to use the current torch device rather than specifying `device_map='sequential'`, it might cause GPU OOM in this way however. Pytests seem to have problems in releasing GPU memory correctly, see https://discuss.pytorch.org/t/torch-pytest-leads-to-memory-fragmentation-how-to-do-proper-integration-testing-of-a-lot-of-torch-models/201231

**Comment by zhenglongjiepheonix** [comment]
> Failure is most probably not related to you, but the slow tests is bad / badly designed as it should either not use accelerate, or not use parallelism to make sure we are testing apples to apples.

Yes, we should never use accelerate, but in order to make tests pass robustly, we might need GPU runners with more memory.

**Review by ArthurZucker — APPROVED** [review]
Wow long due!

**Pull request merged**